packages/react-native/ReactCommon/react/nativemodule/core/platform/android/ReactCommon/JavaTurboModule.cpp

```diff
@@ -391,37 +411,115 @@ jsi::Value createJSRuntimeError(
+                frame->getFileName(),
+                frame->getLineNumber(),
+                frame->getMethodName()
+        });
+    }
+
+    extracted.name = throwable->getClass()->getCanonicalName()->toStdString();
+    extracted.message = throwable->getMessage()->toStdString();
+
+    return extracted;
+}
+
+/**
+ * Converts simple JVM independent Throwable container to JSError.
+ */
+jsi::JSError convertSimpleJThrowableToJSError(
+        jsi::Runtime& runtime,
+        SimpleJThrowable throwable) {
+    auto stackTrace = throwable.stackElements;
+
+    jsi::Array stackElements(runtime, stackTrace.size());
+    for (int i = 0; i < stackTrace.size(); ++i) {
+        auto frame = stackTrace[i];
+
+        jsi::Object frameObject(runtime);
+        frameObject.setProperty(runtime, "className", frame.className);
+        frameObject.setProperty(runtime, "fileName", frame.fileName);
+        frameObject.setProperty(runtime, "lineNumber", frame.lineNumber);
+        frameObject.setProperty(runtime, "methodName", frame.methodName);
+        stackElements.setValueAtIndex(runtime, i, std::move(frameObject));
+    }
+
+    jsi::Object cause(runtime);
+    auto name = throwable.name;
+    auto message = throwable.message;
+    cause.setProperty(runtime, "name", name);
+    cause.setProperty(runtime, "message", message);
+    cause.setProperty(runtime, "stackElements", std::move(stackElements));
+
+    jsi::Value error =
+            createJSRuntimeError(runtime, "Exception in HostFunction: " + message);
+    error.asObject(runtime).setProperty(runtime, "cause", std::move(cause));
+    return jsi::JSError{runtime, std::move(error)};
+}
+
 /**
  * Creates JSError with current JS runtime stack and Throwable stack trace.
  */
 jsi::JSError convertThrowableToJSError(
-    jsi::Runtime& runtime,
-    jni::local_ref<jni::JThrowable> throwable) {
-  auto stackTrace = throwable->getStackTrace();
-
-  jsi::Array stackElements(runtime, stackTrace->size());
-  for (int i = 0; i < stackTrace->size(); ++i) {
-    auto frame = stackTrace->getElement(i);
-
-    jsi::Object frameObject(runtime);
-    frameObject.setProperty(runtime, "className", frame->getClassName());
-    frameObject.setProperty(runtime, "fileName", frame->getFileName());
-    frameObject.setProperty(runtime, "lineNumber", frame->getLineNumber());
-    frameObject.setProperty(runtime, "methodName", frame->getMethodName());
-    stackElements.setValueAtIndex(runtime, i, std::move(frameObject));
-  }
+        jsi::Runtime& runtime,
+        jni::local_ref<jni::JThrowable> throwable) {
+    return convertSimpleJThrowableToJSError(runtime, extractThrowable(throwable));
+}
+
+void rejectWithException(
+    std::optional<AsyncCallback<>> &reject,
+    std::exception_ptr exception,
+    std::optional<std::string> jsInvocationStack) {
+    auto localThrowable = jni::getJavaExceptionForCppException(exception);
+    auto simpleThrowable = extractThrowable(localThrowable);
```

Comment:
Why do we need to extract the throwable here? If we turn `localThrowable` into a global_ref (see jni::make_global), we should be able to pass it into the `call` invocation.

Reply:
https://github.com/facebook/react-native/pull/37484/commits/e99fb6a7ac382ded174c4b97d6054dc4c369330e